### PR TITLE
frogger game

### PR DESCRIPTION
**Pull request by ustymchyk**
### JS OOP

[Demo](https://ustymchyk.github.io/frogger/) |
[Code base](https://github.com/ustymchyk/frogger)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@ustymchyk looks like you have patched the game engine. Original task from Udacity urges to change `app.js` only.

There is a number of matters to consider before intervening with engine code base:

1. Original license may prohibit engine modification.

2. Imagine multiple developers cooperate developing a game, then one of the developers improves the engine everybody else relies upon, while the task is to develop the game and not to improve the engine.

3. With regards to the code review, any changes to the engine would require reviwers to look into it as well. This will take some extra efforts and time. Reviewers may be restricted in their capacity they have to share among multiple students in the course.

Please, restore engine code base and develop the app relying on the engine given with all its flaws and inconsistencies it may have.

**Comment by ustymchyk** [comment]
@OleksiyRudenko 
'Original task from Udacity urges to change app.js only' - I`m not sure about this(proofs in screenshots);

The game can`t be reset expectedly without changes in the engine.
The same with collision, it can be implemented in app.js, but a much better way for this - in the engine(in my opinion).

I understand 1-3 positions, but the course and sources have no information about this.
I can rewrite my code, but, I think this is not my mistake. 

image
image

**Comment by OleksiyRudenko** [comment]
I didn't say you made a mistake, did I?

**Comment by ustymchyk** [comment]
@OleksiyRudenko done
I`ve thought that it will be harder and take more time 😁

File: submissions/ustymchyk/js-oop/app.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+class Enemy {
+    constructor(row, player) {
+        this.sprite = 'images/enemy-bug.png';
+        this.row = row;
+        this.player = player;
+    }
+    init() {
+        this.setSpeed();
+        this.setOffset();
+        this.y = this.row * Helper.cellHeight - 25;
```
Comment (OleksiyRudenko) [inline submissions/ustymchyk/js-oop/app.js:105]: A mysterious number.

File: submissions/ustymchyk/js-oop/app.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        } else {
+            this.x = -Helper.cellWidth;
+            this.setSpeed();
+        }
+
+        this.setCenter();
+        this.checkCollision();
+    }
+    setSpeed() {
+        this.speed = Math.random() * 300 + 200;
```
Comment (OleksiyRudenko) [inline submissions/ustymchyk/js-oop/app.js:121]: Mysterious numbers.

File: submissions/ustymchyk/js-oop/app.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    setCenter() {
+        this.centerX = this.x + Helper.cellWidth / 2;
+        this.centerY = this.y + Helper.cellOffsetTop + Helper.cellHeight / 2;
+    }
+    checkCollision() {
+        const playerCenterX = this.player.centerX;
+        const playerCenterY = this.player.centerY;
+
+        if (Math.abs(playerCenterX - this.centerX) < 65 && Math.abs(playerCenterY - this.centerY) < 55) {
```
Comment (OleksiyRudenko) [inline submissions/ustymchyk/js-oop/app.js:134]: Mysterious number.

**Review by OleksiyRudenko — APPROVED** [review]
@ustymchyk well done!
There are mysterious numbers yet with unclear semantic purpose.
At all times think of how your code **reads**. The closer to plain English the better.

**Pull request merged**